### PR TITLE
EventGrid data spec for Microsoft.Devices/Iothub

### PR DESCRIPTION
**Pull request by cpjagan**
Adding IotHub data schema for event grid specification

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Review by kalyanaj — COMMENTED** [review]
Looks good to me.

**Review by dsgouda — CHANGES_REQUESTED** [review]
Few redundancies that can be reduced by consolidating a few definitions and creating `properties` objects for inline properties.
Also, definitions defined here are not being referenced anywhere, so not sure on the intentions behind this PR

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 1]
```diff
@@ -0,0 +1,274 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "2018-01-01",
+    "title": "Schema of Azure IoT Hub events published to Azure Event Grid",
+    "description":
+      "Describes the schema of the Azure IoT Hub events published to Azure Event Grid. This corresponds to the Data property of an EventGridEvent."
+  },
+  "paths": {},
```
[the remark this answers is not in the document]
Comment (kalyanaj) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: This is used to document the schemas of the various events published to EventGrid. A consumer who is consuming events from EventGrid (e.g. in an Azure function or in any webhook) can use the generated classes to deserialize the event appropriately.  This is the same pattern used for other existing event schemas as well (e.g. the storage events in the eventgrid data plane directory).

[the remark this answers is not in the document]
Comment (dsgouda) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: Thanks for the explanation.

**Comment by cpjagan** [comment]
@dsgouda I have pushed changes based on your comments to split into smaller definitions.

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+            },
+            "properties": {
+              "type": "object",
+              "description": "Properties JSON element.",
+              "properties": {
+                "desired": {
+                  "type": "object",
+                  "description":
+                    "A portion of the properties that can be written only by the application back-end, and read by the device.",
+                  "properties": {
```
[the remark this answers is not in the document]
Comment (cpjagan) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: Yes, the model has properties with "$" to signify that they are system properties.

[the remark this answers is not in the document]
Comment (dsgouda) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: If these are not odata properties, we strongly recommend using `x-ms-client-name` ([here](https://github.com/Azure/azure-rest-api-specs/blob/0bef2b99ad6d52781e3c308bc6b420f6eef59ef7/documentation/openapi-authoring-automated-guidelines.md#L351))
You can name the property without special characters and specify its exact name with the special character as the `x-ms-client-name` value

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    "version": "2018-01-01",
+    "title": "Schema of Azure IoT Hub events published to Azure Event Grid",
+    "description":
+      "Describes the schema of the Azure IoT Hub events published to Azure Event Grid. This corresponds to the Data property of an EventGridEvent."
+  },
+  "paths": {},
+  "definitions": {
+    "IotHubDeviceCreatedEventData": {
+      "description": "Schema of the Data property of an EventGridEvent for an Microsoft.Devices.DeviceCreated event.",
+      "$ref" : "#/definitions/DeviceLifeCycleEventData"
```
Comment (dsgouda) [inline specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: The right way to do this is:
```json
"IotHubDeviceCreatedEventData": {
    "properties":{
          "x-ms-client-flatten": true,
          "$ref": "#/definitions/DeviceLifeCycleEventData",
          "description": "device life cycle properties."
   }
}
```

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    },
+    "IotHubDeviceDeletedEventData": {
+      "description": "Schema of the Data property of an EventGridEvent for an Microsoft.Devices.DeviceDeleted event.",
+      "$ref" : "#/definitions/DeviceLifeCycleEventData"
+    },
+    "DeviceLifeCycleEventData": {
```
Comment (dsgouda) [inline specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: A better name here is `DeviceLifeCycleEventProperties`

**Review by dsgouda — CHANGES_REQUESTED** [review]

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json (hunk 3 above)
Comment (cpjagan) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: I would have to use the `"x-ms-client-flatten": true,` everywhere I reference is it?
```

**Comment by cpjagan** [comment]
@dsgouda Pushed fixes based on your comment. Please review and let me know.

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  "info": {
+    "version": "2018-01-01",
+    "title": "Schema of Azure IoT Hub events published to Azure Event Grid",
+    "description":
+      "Describes the schema of the Azure IoT Hub events published to Azure Event Grid. This corresponds to the Data property of an EventGridEvent."
+  },
+  "paths": {},
+  "definitions": {
+    "IotHubDeviceCreatedEventData": {
+      "description": "Schema of the Data property of an EventGridEvent for a Microsoft.Devices.DeviceCreated event.",
```
Comment (dsgouda) [inline specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: nit: A better description would be 
"Event data for IotHubDeviceCreated event"

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            },
+            "secondaryThumbprint": {
+              "type": "string",
+              "description": "Secondary thumbprint for the x509 certificate."
+            }
+          }
+        }
+      }
+    },
+    "DeviceTwinProperties": {
```
Comment (dsgouda) [inline specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: nit: Use a more meaningful name here, there already exists a definition for `DeviceTwinInfo` with some properties
something like `TwinDeviceReportedProperties` or `DeviceReportedProperties` based on the property name

**Review by dsgouda — CHANGES_REQUESTED** [review]
A couple of minor comments and still recommend usage of `x-ms-client-name` for property names with special characters, e.g." 
```json
 "lastUpdated": {
          "type": "string",
          "description": "The ISO8601 timestamp of the last time the properties were updated.",
          "x-ms-client-name":"$lastUpdated"
        }

```

File: specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json (hunk 6 above)
Comment (cpjagan) [reply specification/eventgrid/data-plane/Microsoft.Devices/stable/2018-01-01/IotHub.json]: in this context reported properties has a different meaning. So using that would only confuse the reader who understands the iot hub twin concepts. DeviceTwinProperty is the right name.

Also, TwinInfo and TwinProperties were names I picked from another swagger doc that a partner team of us already uses. I would like to be consistent with those. 

I will pass on this nit.

**Review by dsgouda — APPROVED** [review]
LGTM pending CI builds

**Pull request merged**